app/src/main/java/com/github/bgabriel998/softwaredevproject/gallery/ZoomableImageView.java

```diff
@@ -0,0 +1,311 @@
+                    firstTouch(currTouch);
+                    break;
+
+                case MotionEvent.ACTION_MOVE :
+                    drag(currTouch);
+                    break;
+
+                case MotionEvent.ACTION_UP :
+                    mode = Mode.None;
+                    int xDiff = (int) Math.abs(currTouch.x - start.x);
+                    int yDiff = (int) Math.abs(currTouch.y - start.y);
+                    if (xDiff < CLICK_DISTANCE_THRESH &&
+                            yDiff < CLICK_DISTANCE_THRESH) {
+                        v.performClick();
+                    }
+                    break;
+
+                case MotionEvent.ACTION_POINTER_UP :
+                    mode = Mode.None;
+                    break;
+            }
+
+            setImageMatrix(imageMatrix);
+            invalidate();
+            return true;
+        });
+    }
+
+    /**
+     * Called on first touch
+     * Sets mode and touch positions.
+     * @param currTouch the current/first touch
+     */
+    private void firstTouch(PointF currTouch) {
+        start.set(currTouch);
+        last.set(currTouch);
+        mode = Mode.Drag;
+    }
+
+    /**
+     * Preform a drag of a touch, by translating the matrix
+     * @param currTouch current touch position
+     */
+    private void drag(PointF currTouch) {
+        if (mode == Mode.Drag) {
+            float deltaX = (viewWidth < currWidth) ? currTouch.x - last.x : 0;
+            float deltaY = (viewHeight < currHeight) ? currTouch.y - last.y : 0;
+            imageMatrix.postTranslate(deltaX, deltaY);
+            fixMatrixTranslation();
+            last.set(currTouch);
+        }
+    }
+
+    /**
+     * Fix the matrix Translation
+     */
+    private void fixMatrixTranslation() {
+        float[] m = new float[9];
+        imageMatrix.getValues(m);
+        float transX = m[Matrix.MTRANS_X];
+        float transY = m[Matrix.MTRANS_Y];
+
+        float fixedTransX = getFixTrans(transX, viewWidth, currWidth);
+        float fixedTransY = getFixTrans(transY, viewHeight, currHeight);
+        if (fixedTransX != 0 || fixedTransY != 0) {
+            imageMatrix.postTranslate(fixedTransX, fixedTransY);
+        }
+    }
+
+    /**
+     * Based on a trans point, get the fixed point
+     * @param trans given trans point
+     * @param viewSize the view size
+     * @param currSize the current scaled size
+     * @return the trans point fixed
+     */
+    private float getFixTrans(float trans, float viewSize, float currSize) {
+        float minTrans, maxTrans;
+
+        if (currSize <= viewSize) {
+            minTrans = 0;
+            maxTrans = viewSize - currSize;
+        }
+        else {
+            minTrans = viewSize - currSize;
+            maxTrans = 0;
+        }
+
+        if (trans < minTrans) {
+            return minTrans - trans;
+        }
+        if (trans > maxTrans) {
+            return maxTrans - trans;
+        }
+        return 0;
```

Comment:
```suggestion
        return trans < minTrans ? minTrans - trans : maxTrans - trans;
```

Reply:
Not sure this is correct. I am pretty sure that there exists cases where (for easy numbers) minTrans = 0, maxTrans = 10, trans = 7. Then we want to return 0. not 3

Reply:
Oh yep you're right